### PR TITLE
add signed church integers

### PR DESCRIPTION
**Pull request by billpmurphy**
Adds [signed Church-encoded numbers](https://en.wikipedia.org/wiki/Church_encoding#Signed_numbers).

There are a few things that I'm not so sure about:
* The function names could be simplified - `signed_add` could arguably be changed to `add` given that it is in a different module than `num::church:add`; I don't have an opinion either way
* In `tests/signed.rs`, most of the tests can abstracted out with a macro to make it easier to support other encodings; I'm not sure what the best way to do this is yet
* It's easy to add a Trait that provides an `into_church` method for `i32`. However, this is very misleading, because the expression `1.into_church()` will have an inferred a type of `i32` for `1`, converting it to a signed `Term`, but the probably expects it to be converted to an (unsigned) Church numeral. Maybe a new trait with a method `into_church_signed()` is the way to go here?
* The wording in the documentation for `negate`, `signed_add`, etc can probably be improved.

### CODE REVIEW

**Comment by ljedrz** [comment]
I have a few suggestions:
- since I don't recall ever seeing non-Church signed numerals in the wild (correct me if I'm missing something), it might be a good idea to just put them in a `num::signed` module
- I prefer shorter function names; if different variants of e.g. `add` are needed, `signed::add` will do fine; it's also the convention in other modules
- don't worry about test macros; I intend to improve the other test macros in time, too, but it's not a priority (as long as the functions are well tested regardless of test code repetition)
- I think `into_church_signed()` will be the desired option when some other signed numerals appear; if you don't expect to add them, `into_signed()` will do just fine
- as for the wording, `negate` could probably be changed to `neg`; I'd also change the names of `normalize` (so it's not mixed with `beta`; I'm not sure to what, though), `absolute_value` to `modulus`
- as mentioned in the other PR, since 2.0 I intend for the definitions to match the docs (simplicity+correctness > performance); if you still have the more basic definitions, I'd prefer if you rolled back to them

I have a feeling that this notation for signed numerals could be extended to other encodings too; maybe you could investigate adding an `Encoding` enum e.g. in `num::convert` and allowing e.g. `signed::add(Pargiot)`?

**Comment by billpmurphy** [comment]
Updated; I went for the simpler option and restricted signed numerals to Church numerals, but I think it's relatively easy to refactor to be more generic.

File: src/data/num/convert.rs [hunk 1]
```diff
@@ -87,6 +88,18 @@ impl IntoBinaryNum for usize {
         abs!(3, ret)
     }
 }
+
+impl IntoSignedNum for i32 {
+    fn into_signed(self) -> Term {
+        let numeral = (self.abs() as usize).into_church();
+        if self > 0 {
+            abs(app!(Var(1), numeral, abs!(2, Var(1))))
```
Comment (ljedrz) [inline src/data/num/convert.rs]: you can use the `tuple!()` macro here

File: src/data/num/convert.rs [hunk 2]
```diff
@@ -87,6 +88,18 @@ impl IntoBinaryNum for usize {
...
+        } else {
+            abs(app!(Var(1), abs!(2, Var(1)), numeral))
```
Comment (ljedrz) [inline src/data/num/convert.rs]: and here

File: src/data/num/signed.rs [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+            )),
+            I()
+        ))
+    )
+}
+
+/// Applied to a pair representing a signed Church-encoded integer it returns its absolute
+/// value as a Church numeral.
+///
+/// ABSOLUTE_VALUE ≡ λx.(λy.IS_ZERO (FST y) (SND y) (FST y)) (NORMALIZE x) ≡
```
Comment (ljedrz) [inline src/data/num/signed.rs]: ABSOLUTE_VALUE > MODULUS

Comment (ljedrz) [reply src/data/num/signed.rs]: NORMALIZE > SIMPLIFY

File: src/data/num/signed.rs [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+/// λ (λ IS_ZERO (FST 1) (SND 1) (FST 1)) (NORMALIZE 1)
```
Comment (ljedrz) [inline src/data/num/signed.rs]: NORMALIZE > SIMPLIFY

File: src/data/num/signed.rs [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+/// ```
+pub fn modulus() -> Term {
+    abs(app(
+        abs(app!(
+            church::is_zero(),
+            app(fst(), Var(1)),
+            app(snd(), Var(1)),
+            app(fst(), Var(1))
+        )),
+        app(simplify(), Var(1))
```
Comment (ljedrz) [inline src/data/num/signed.rs]: Is this `simplify()` necessary to return valid solutions (if it is then fine)?

File: src/data/num/signed.rs [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+    ))
+}
+
+/// Applied to a pair of two Church-encoded numerals representing a signed integer it returns a
+/// pair representing their sum.
+///
+/// ADD ≡ λa.λb.NORMALIZE (PAIR (ADD (FST a) (FST b)) (ADD (SND a) (SND b))) ≡
+/// λ λ NORMALIZE (PAIR (ADD (FST 2) (FST 1)) (ADD (SND 2) (SND 1)))
```
Comment (ljedrz) [inline src/data/num/signed.rs]: in this case I'm pretty sure we can leave the result unsimplified, that's the convention in other modules too

File: src/data/num/signed.rs [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+            )
+        )
+    ))
+}
+
+/// Applied to a pair of two Church-encoded numerals representing a signed integer it returns a
+/// pair representing their difference.
+///
+/// SUB ≡ λa.λb.NORMALIZE (PAIR (ADD (FST a) (SND b)) (ADD (SND a) (FST b))) ≡
+/// λ λ NORMALIZE (PAIR (ADD (FST 2) (SND 1)) (ADD (SND 2) (FST 1)))
```
Comment (ljedrz) [inline src/data/num/signed.rs]: in this case I'm pretty sure we can leave the result unsimplified too

File: src/data/num/signed.rs (hunk 5 above)
Comment (billpmurphy) [reply src/data/num/signed.rs]: Yes, here it's required; it's not required in `add`/`sub`

**Comment by billpmurphy** [comment]
Updated - I think there is a case for keeping `simplify` in `add` and `sub`:

* It matches the definitions given on wikipedia
* It allows the equalities like the ones in the tests to hold, where you can call `sub` and `add` then compare the result to another signed number
* it makes it impossible to create a non-simplified signed numeral (assuming you don't manipulate the pair directly). 

However, none of these are very strong reasons - I'll make the change if you'd rather remove `simplify` from `add` and `sub`.

**Pull request merged**